fenicsxconcrete/experimental_setup/concrete_cylinder.py

```diff
@@ -0,0 +1,251 @@
+        p = Parameters()
+        
+
+        p['degree'] = 2 * ureg('')  # polynomial degree
+
+        p.update(parameters)
+
+        super().__init__(p)
+
+        # initialize variable top_displacement
+        self.top_displacement = df.fem.Constant(domain=self.mesh, c=0.0)  # applied via fkt: apply_displ_load(...)
+
+    def setup(self):
+        """Generates the mesh based on parameters
+
+        This function is called during __init__
+        """
+
+        if self.p['dim'] == 2:
+            # build a rectangular mesh to approximate a 2D cylinder
+            self.mesh = df.mesh.create_rectangle( MPI.COMM_WORLD, 
+                                                 [[0.0, 0.0],[self.p['radius']* 2, self.p['height']],],
+                                                 [self.p['mesh_density'], self.p['mesh_density']],
+                                                 cell_type=df.mesh.CellType.triangle,)
+        elif self.p['dim'] == 3:
+            # generates a 3D cylinder mesh based on radius and height
+            # to reduce approximation errors due to the linear tetrahedron mesh, the mesh radius is iteratively changed
+            # until the bottom surface area matches that of a circle with the initially defined radius
+            def create_cylinder_mesh(radius, p):
+                # generate cylinder mesh using gmsh
+                mesh = generate_cylinder_mesh(radius, p['height'], p['mesh_density'])
+                facets = df.mesh.locate_entities_boundary(mesh, 2, plane_at(0.0, 2))
+                tdim = mesh.topology.dim
+                f_v = mesh.topology.connectivity(tdim - 1, 0).array.reshape(-1, 3)
+                entities = df.graph.create_adjacencylist(f_v[facets])
+                values = np.full(facets.shape[0], 2, dtype=np.int32)
+
+                ft = df.mesh.meshtags_from_entities(mesh, tdim - 1, entities, values)
+                ds = ufl.Measure("ds", domain=mesh, subdomain_data=ft)
+                bottom_area = df.fem.assemble_scalar(df.fem.form(1 * ds(2)))
+
+                return bottom_area, mesh
+
+            # create a discretized cylinder mesh with the same cross-sectional area as the round cylinder
+            target_area = np.pi * self.p['radius']**2
+            effective_radius = self.p['radius']
+            mesh_area = 0 
+            area_error = 1e-6
+            #
+            # iteratively improve the radius of the mesh till the bottom area matches the target
+            while abs(target_area - mesh_area) > target_area * area_error:
+                # generate mesh
+                self.p["mesh_radius"] = effective_radius  # not required, but maybe interesting as metadata
+                mesh_area, self.mesh = create_cylinder_mesh(effective_radius, self.p)
+                # new guess
+                effective_radius = np.sqrt(target_area / mesh_area) * effective_radius
+
+        else:
+            raise Exception(f"wrong dimension {self.p['dim']} for problem setup")
```

Comment:
this line is not tested, please add a test for "wrong dimension". Easiest would be to change Exception to ValueError and add concrete_cylinder to the lists in:
https://github.com/BAMresearch/FenicsXConcrete/blob/37-improve-coverage-for-initial-classes/tests/experimental_setup/test_experimental_setups.py
this change is currently not in main but you can either review #41 or just copy the changes from branch #37

Reply:
Added tests